ouroboros-network/src/Ouroboros/Network/Protocol/BlockFetch/Server.hs

```diff
@@ -189,23 +190,37 @@ blockFetchServerSenderToProducer (BlockFetchServerSender mserver) =
     join $ Pipes.lift (streamBlocks <$> mstreamer)
   streamBlocks (SendMessageBatchDone sender) = blockFetchServerSenderToProducer sender
 
+-- | Sender of the @'BlockFetchServerProtocol'@.  It may sand
+-- @'MessageServerDone'@ under two conditions:
+--
+--  * @'RangeStream' range@ returned @'End'@, which means that the
+--    corresponding @'BlockFetchClientProtocol'@ has terminated
+--  * the @'blockStream'@ handler returned @'Nothing'@
+--
 blockFetchServerSender
   :: forall m range block a.
      Monad m
   => a
-  -> m range
+  -- ^ return value
+  -> m (RangeStream range)
+  -- ^ stream of ranges; when @'End'@ is read the protocol will terminate.
   -> (range -> m (Maybe (Pipes.Producer block m a)))
+  -- ^ range handler returning a stream of blocks, when it return @'Nothing'@
+  -- the @'MessageServerDone'@ will terminate the protocol.
```

Comment:
Could we instead use `Pipes.Producer (Pipes.Producer block m ()) m a`?

Reply:
We could:  when entering  an inner `Producer` we send `MessageStartBatch`, then
* if empty we send `MessageNoBlocks`
* if non-empty we send `MessageBlock`s followed by `MessageBatchDone` 
* and loop until the outer Producer is exhausted.

If the outer one was empty we just send `MessageNoBlocks`.

The only problem I see with this, is that the `MessageNoBlocks` is used in two different ways.  The client will have a problem to distinguish a producer which send `MessageNoBlocks` which then streams some blocks with a one that responds with `MessageNoBlocks` since the outer producer did not contain any blocks.

But I think one can map `Producer (Producer block m ()) m a -> m (Maybe (Producer m a)`. It will force first element of the outer `Producer` to check if we need `Just` or `Nothing`, and also the first element of the first inner producer.

`Maybe (Producer block m a)` seems cleaner to me.

Reply:
Oh, the protocol seems a little too complicated then. Why not simply send, in response to a request, a message for each block, then a message saying the batch is done? Why do we need the extra complexity of "start batch" and "no blocks"?

Reply:
This way the server can handle multiple requests from a client, each one in a separate 'batch', and if there's no block for a given request the client will receive `MessageNoBlocks`. Initially, I had a message that started a batch together with sending a block, but @dcoutts preferred to have a separate message for starting a batch.

Reply:
We could simplify this further and use a single protocol in both ways: one streaming protocol.  It's probably  better to keep them separate since we want to have different error conditions (for streaming blocks we need `MessageNoBlocks`, but we don't need such message for sending requests of ranges).

Reply:
@avieth block fetch logic will maintain this [invariant](https://github.com/input-output-hk/ouroboros-network/pull/101/files#diff-08a38dc47901023f6ade5e0182b57bb5R251):

>  We maintain and rely on the invariant that the ranges of fetched blocks are
 backwards closed. This means we never have discontinuous ranges of fetched or
 not-yet-fetched blocks.

It will be simpler to maintain it with the `MessageStartBatch` and `MessageBatchDone` since the client will be able to associate requested ranges with received blocks. After receiving a range of blocks the client could compute received range and see if it received what it was requesting and handle the blocks to the consensus layer.  This is still possible without these two messages, but this way it's more explicit and overhead is small (single round trip).

Reply:
Oh I think I understand. Since this protocol includes no client requests (no states are in client partition), there's no way to know from _within_ the protocol which block message is associated with which batch request?

Reply:
The multi-part response protocol would be simpler, IMO.

```Haskell
data StMPR where
  StIdle :: StMPR
  StBusy :: StMPR
  StDone :: StMPR

data TrMPR request response from to where
  TrRequest :: request -> TrMPR request response 'StIdle 'StBusy
  TrRespondPart :: response -> TrMPR request response 'StBusy 'StBusy
  TrRespondDone :: TrMPR request response 'StBusy 'StIdle
  TrDone :: TrMPR request response 'StIdle 'StDone

type TrBlockDownload = TrMPR BlockRange BlockBody
```

Have 2 threads leap-frog each-other, each running that protocol and checking the batching invariants _locally_, within the protocol application itself (which is very easy).

Reply:
^ I also think this will be less puzzling to other readers. Two protocols, each with no client or no server states, running completely independently, yet linked by way of shared mutable state and subject to nontrivial invariants, is just confusing.

Reply:
One will not achieve pipelining with this approach, what @duncan wanted to achieve. It may be possible to extend the typed framework so that we can have the total `connect` function when running two dual protocols against each other concurrently.

What you get here are two threads that are untagged, no synchronisation is needed. 

One can interpret the client side of `BlickFetchClientProtocol`as a `Pipes.Producer`, then it's server side as a `Pipes.Pipe` which feeds into the server side of a `BlockFetchServerSender` which is another `Pipes.Pipe` which feeds into the receiver side which is isomorphic to a `Pipes.Consumer`. That's a one producer, two pipes and a consumer: quite clear design.

A pipelining protocol will have this kind of state machine description: a product of two dual state machines.